src/box/space_upgrade.c

```diff
@@ -18,23 +18,48 @@
 # error unimplemented
 #endif
 
+#define BAD_PTR ((void *)-1)
+
 struct space_upgrade_def *
 space_upgrade_def_decode(const char **data, struct region *region)
 {
-	(void)data;
 	(void)region;
-	diag_set(ClientError, ER_UNSUPPORTED,
-		 "Community edition", "space upgrade");
+	/**
+	 * Option decoder may only fail with IllegalParams error so we return
+	 * a non-NULL ptr here and abort later, in space_upgrade_check_alter.
+	 */
+	mp_next(data);
+	return BAD_PTR;
+}
+
+struct space_upgrade_def *
+space_upgrade_def_dup(const struct space_upgrade_def *def)
+{
+	if (def == BAD_PTR)
```

Comment:
I'm not sure if it's legal in modern C.

[C11 § 6.5.9 paragraph 6](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1548.pdf) (pointed by @drewdzzz):
> Two pointers compare equal if and only if both are null pointers, both are pointers to the same object (including a pointer to an object and a subobject at its beginning) or function, both are pointers to one past the last element of the same array object, or one is a pointer to one past the end of one array object and the other is a pointer to the start of a different array object that happens to immediately follow the first array object in the address space.)

Reply:
Sounds fair. Replaced `BAD_PTR` with `dummy_def`.

<details><summary>diff</summary>

```diff
diff --git a/src/box/space_upgrade.c b/src/box/space_upgrade.c
index 3c3fe6b42c82..6d6b8b59584b 100644
--- a/src/box/space_upgrade.c
+++ b/src/box/space_upgrade.c
@@ -18,7 +18,7 @@
 # error unimplemented
 #endif
 
-#define BAD_PTR ((void *)-1)
+static struct space_upgrade_def {} dummy_def;
 
 struct space_upgrade_def *
 space_upgrade_def_decode(const char **data, struct region *region)
@@ -29,22 +29,20 @@ space_upgrade_def_decode(const char **data, struct region *region)
 	 * a non-NULL ptr here and abort later, in space_upgrade_check_alter.
 	 */
 	mp_next(data);
-	return BAD_PTR;
+	return &dummy_def;
 }
 
 struct space_upgrade_def *
 space_upgrade_def_dup(const struct space_upgrade_def *def)
 {
-	if (def == BAD_PTR)
-		return BAD_PTR;
-	assert(def == NULL);
-	return NULL;
+	assert(def == NULL || def == &dummy_def);
+	return def != NULL ? &dummy_def : NULL;
 }
 
 void
 space_upgrade_def_delete(struct space_upgrade_def *def)
 {
-	assert(def == NULL || def == BAD_PTR);
+	assert(def == NULL || def == &dummy_def);
 	(void)def;
 }
 
@@ -53,7 +51,7 @@ space_upgrade_check_alter(struct space *space, struct space_def *new_def)
 {
 	assert(space->upgrade == NULL);
 	assert(new_def->opts.upgrade_def == NULL ||
-	       new_def->opts.upgrade_def == BAD_PTR);
+	       new_def->opts.upgrade_def == &dummy_def);
 	(void)space;
 	if (new_def->opts.upgrade_def != NULL) {
 		diag_set(ClientError, ER_UNSUPPORTED,
```
</details>